### ISSUE TITLE
How to implement a custom Keras loss function, which uses an external routine

### ISSUE THREAD

**Issue by jasiozaucha** [fchollet/keras#6652]
I am trying to define a custom loss, which is implemented in an external routine (sometimes not even python). 

Essentially I would like to extract (y_true, y_pred) as numpy arrays to feed into my routine, which will return a value that I can then feed back to Keras. Is this possible? 

`
from keras import backend as K

    def _loss(y_true,y_pred):
        tf_session = K.get_session()
        y_true_matrix = y_true.eval(session = tf_session)
        y_pred_matrix = y_pred.eval(session = tf_session)
        distance = external_routine(y_true_matrix, y_pred_matrix)
        return K._to_tensor(distance, dtype=DT_FLOAT)
`

**Comment by Dref360** [comment]
The loss function is part of the graph. So you need to do everything using Tensors.

**Comment by jasiozaucha** [comment]
I know that's the standard way, but the routine is much too complicated. I think my network would benefit from this custom routine, it is very specific to phylogenomics, I can't implement it all using tensors... 
Are you saying it's impossible to do what I am trying?

**Comment by bfarzin** [comment]
It is not possible.  For back propagation to work, you need to take the gradient of the loss w.r.t. the parameters.  The Keras backend (Tensorflow or Theano) take the gradient for you in the background using the graph for the model and the loss function.  See [here](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L1209). If the function is not differential in the backend, then it cannot propagate the errors for the next step of your optimizer.

**Comment by AlexImmer** [comment]
It depends on the computations you cannot do within tensorflow. If you can express your custom routines as a change in y_true for example, it is possible. In fact, you have to ensure that you can still compute a loss in tensorflow/theano/cntk which can be derived to update your model.

For a custom routine defined in pure python, have a look here: [tf.py_func](https://www.tensorflow.org/api_docs/python/tf/py_func)
